File: fsharp/bob/Bob.fs
module Bob
open System
open System.Text.RegularExpressions

let isUppercaseChar char =
    if(Char.IsLetter char)
        then Char.IsUpper char |> Some
        else None

let isUppercase string =
    let charSeq = Seq.toList string
    if(List.exists Char.IsLetter charSeq)
    then charSeq 
        |> List.filter Char.IsLetter
        |> List.forall Char.IsUpper
    else
        false

let stringEndsWith (string: string) char =
    string.Chars(string.Length-1) = char

let response string =
    let m = Regex.Match(string, "^(.*?)\s*$")
    if(m.Success)
    then
        let string = m.Groups.[1].Value
        if(Regex.IsMatch(string, "^\s*$"))
        then "Fine. Be that way!"
        else
            if(isUppercase string)
            then
                if(stringEndsWith string '?')
                then "Calm down, I know what I'm doing!"
                else "Whoa, chill out!"
            else
                if(stringEndsWith string '?')
                then "Sure."
                else "Whatever."
    else
        "Whatever."

File: csharp/bob/Bob.cs
using System.Linq;

public static class Bob
{
    public static string Response(string statement)
    {
        switch (statement.Trim())
        {
            case var s when IsShouting(s) && IsQuestion(s):
                return "Calm down, I know what I'm doing!";
            case var s when IsShouting(s):
                return "Whoa, chill out!";
            case var s when IsQuestion(s):
                return "Sure.";
            case var s when IsSilent(s):
                return "Fine. Be that way!";
            default:
                return "Whatever.";
        }
    }

    private static bool IsShouting(string s)
    {
        var ca = s.ToCharArray();
        return ca.Any(char.IsUpper) && ca.All(c => !char.IsLower(c));
    }

    private static bool IsQuestion(string s) => s.EndsWith("?");

    private static bool IsSilent(string s) => string.IsNullOrEmpty(s);
}